Arch arm mpu re work

ioannisg opened the pull request:
Pull-Request for the MPU support for ARM.

Fixes #12809 re-working the ARM MPU architecture

Besides that, the PR does the following:

Fixes #8907 
Fixes #12548
Fixes #12821 (since `configure_mpu_stack_guard()` is depreciated)
Fixes #12688 (by implementation. it unifies behavior on ARM and NXP)

Closes #10633 
Closes #12822 
Addresses the Arm side of #12555

---

Verification: tested the implementation with:
- the whole `tests/kernel/mem_protect` test suite for all MPU architectures
- `samples\mpu\mpu_stack_guard_test` for all MPU architectures.

Testing boards:
   - `nrf52_pca10040` for ARMv7-M
   - `frdm_k64f` for NXP MPU
   - `nrf9160_pca10090` for ARMv8-M
   - `sam_e70_xplained` for ARMv7-M

ioannisg committed:
arch: arm: Kconfig option for non-overlapping MPU region requirement

This commit introduces a new Kconfig option symbol,
MPU_REQUIRES_NON_OVERLAPPING_REGIONS, to signify the use of an
MPU architecture that requires the active MPU regions be
non-overlapping. This requirement concerns the standard
(unmodified) ARMv8-M MPU implementation. For that, we select
the option in the definition of ARM_MPU, if the CPU architecture
is ARMv8-M (Baseline or Mainline).

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
kernel: mem_domain: extend sane_partition for non-overlapping regions

This commit extends the implementation of sane_partition(..) in
kernel/mem_domain.c so that it generates an ASSERT if partitions
inside a mem_domain overlap. This extension is only implemented
for the case when the MPU requires non-overlapping regions.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: introduce/implement API for configuring static MPU regions

This commit introduces the generic ARM (core) API, which allows
the user to program a set of static (fixed) MPU regions at boot
time. The API function is invoked upon initialization, in the
ARM-specific call of _arch_switch_to_main_thread(). The API
implementation is provided in arm_core_mpu.c.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: introduce/implement API to configure dynamic MPU regions

This commit introduces an ARM API that allows the user to
program a set of dynamic MPU regions at run-time. The API
function is invoked every time the memory map needs to be
re-programmed (for example at thread context-switch). The
functionality is implementated in arm_core_mpu.c.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: mpu: internal API for configuring static memory regions

This commit introduces and documents the internal ARM MPU
API to configure the static memory regions at boot time.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: mpu: implement internal API for configuring static regions

This commit implements the common ARMv7-M and ARMv8-M internal
shim for configuring static memory regions at boot time. It also
adds LOG error messages, and ASSERTS, which fire if the static
MPU region configuration is not performed successfully.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: mpu: remove init of static regions from arm_mpu_init()

This commit removes the initialization of static memory regions
from arm_mpu_init(), since they are now initialized, directly,
by the kernel. The programming of MPU regions configured in SOC
definition is kept here.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: mpu: implement API for memory domains in ARM (core) MPU

This commit updates the ARM Core MPU API for memory domains,
to align with the principle of de-coupling the partitioning
and the access attribution with the architecture-specific
MPU driver implementation.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: mpu: functionality to configure static memory regions (v7)

This commit contains the added functionality to program the
ARM7-M MPU, so it can configure the static memory regions at
boot time.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: mpu: functionality to configure static memory regions (v8)

This commit contains the added functionality to program the
ARM8-M MPU, so it can configure the static memory regions at
boot time, as well as mark the areas, inside which dynamic MPU
region configuration at run-time will be allowed.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: mpu: internal API for configuring dynamic memory regions

This commit introduces and documents the internal ARM MPU
API to configure the dynamic memory regions at run-time.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: mpu: implement internal API to configure dynamic regions

This commit implements the common ARMv7-M and ARMv8-M internal
shim for configuring dynamic memory regions at run-time.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: mpu: internal API for re-configuring active partitions

This commit introduces an internal ARM MPU API that allows the
user to re-configure a memory partition in run-time.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: mpu: implement arm_core_mpu_get_max_available_dyn_regions

This commit adds the implementation of internal ARM MPU
function, arm_core_mpu_get_max_available_dyn_regions(),
required by _arch_mem_domain_max_partitions_get(),
according to the new architecture. The function returns
the number of the available partitions for memory domains
based on the total amount of HW regions and the
actual number of statically configured MPU regions.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: mpu: functionality to configure dynamic regions (v7)

This commit contains the added functionality to program the
ARM7-M MPU so it can configure the dynamic memory regions at
run-time.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: mpu: implement internal API to configure partitions

This commit implements the internal (ARMv7-M and ARMv8-M)
API, which allows to user to (re)configure a memory partition
at run-time.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: mpu: group together USERSPACE-related implementation

This commit moves all internal implementation, relevant for
User mode, in a single place in the arm_mpu_v7_internal.h,
arm_mpu_v8_internal, and in arm_mpu.c. Additionally, the
commit cleans up internal function _get_region_attr() that
is not used any more.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: clean up depreciated implementation and API definition

This commit removes obsolete ARM CORE MPU API definitions
and related implementation from arm_mpu.c, in the wake of
the transition to the new ARM MPU design.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: mpu: functionality to configure dynamic regions (v8)

This commit contains the added functionality to program the
ARM8-M MPU so it can configure the dynamic memory regions at
run-time.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: mpu: align NXP MPU implementation with ARM MPU

This commit re-works the NXP MPU driver implementation so that
it aligns with the implementation for ARMv7-M and ARMv8-M MPU
architectures.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: mpu: remove unnecessary MPU region type definitions

This commit removes the unnecessary MPU region type definitions
from arm_core_mpu_dev.h, as they are not used any more in any of
the architecture-specific MPU implementations (ARMv7-M, NXP, and
ARMv8-M MPU).

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
FYI @carlescufi @ruuddw

ioannisg commented:
@aurel32 in principle, with this re-work, **we do not disable the MPU during re-programming**, any more. Shall we claim this PR fixes #12548  ?

andyross commented on arch/arm/core/swap_helper.S:
```diff
@@ -182,13 +182,14 @@ _thread_irq_disabled:
+    ldr r0, =_kernel
+    ldr r0, [r0, #_kernel_offset_to_current]
+    bl _arch_configure_dynamic_mpu_regions
+    pop {r0,r1,r2,r3,lr}
```
Not my business to nit over ARM assembly, obviously, but this looks like you're doing too much work.  The earlier code seems to know that it has the thread pointer in r2 already (i.e. it doesn't have to fetch it from kernel._current) and that r0/r1/r3 are already scratch and don't need to be saved.

I'd look carefully to see if you really need to be doing all this; context switching is the hottest of hot paths.

andyross commented on kernel/mem_domain.c:
```diff
@@ -42,6 +43,13 @@ static bool sane_partition(const struct k_mem_partition *part,
+		__ASSERT(false, "overlapping partitions <%x...%x>, <%x...%x>",
+			part->start, last,
+			parts[i].start, cur_last);
+		return false;
```
Same pattern.  Log an error, don't assert for a runtime condition you expect to be handled.

ioannisg replied:
Oh wait. kernel/mem_domain.c does not use the Logger yet, so I can not leave a LOG_ERR message. Although I totally agree with no having ASSERTS when we do return error signals, I prefer to stick with the rule in this module (and in the very same functio actually) and leave the ASSERT.

We can definitely fill-in an issue to make mem_domain use Logger and also remove ASSERTS entirely when signal errors are propagated.
I will track that externally.

andyross replied:
Then change this to an _ASSERT(!IS_ENABLED(CONFIG_MPU_REQUIRES_NON_OVERLAPPING_REGIONS), "") maybe?  That would better express this as a configuration error and not a runtime one.

Or just printk, honestly.  No shame in printk.

andrewboie replied:
I concur with Andy

aurel32 replied:
Am I correct that the existing code only checks for consecutive overlapping memory partitions?

ioannisg replied:
No, @aurel32 it checks whether the current partition overlaps with any other partition in this domain.

aurel32 replied:
Hmm correct, sorry about the confusion.

andyross reviewed (changes requested):
Lots and lots of nitpickery, most of which can be ignored if you don't like the suggestions (but please do look carefully at that code in _Swap() or have @agross-linaro check it out -- it looks needlessly slow to me).

The only thing I really care about is the pattern you have where you use __ASSERT(0, ...) to get logging for a **runtime** error that you expect production code to handle.  That's really sort of an abuse and leads to bugs where apps wind up with untested error handling code because it never reached it in the development builds.  Please use LOG_ERR() for those.

ioannisg commented:
> 
> 
> Lots and lots of nitpickery, most of which can be ignored if you don't like the suggestions (but please do look carefully at that code in _Swap() or have @agross-linaro check it out -- it looks needlessly slow to me).
> 
> The only thing I really care about is the pattern you have where you use __ASSERT(0, ...) to get logging for a **runtime** error that you expect production code to handle. That's really sort of an abuse and leads to bugs where apps wind up with untested error handling code because it never reached it in the development builds. Please use LOG_ERR() for those.

Thanks for the quick feedback, @andyross , I will address your comments. 
About swap; sure, I would appreciate @agross-linaro to look at it carefully and, yes, we can make sure it runs more efficiently.

pizi-nordic commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -49,6 +45,58 @@ static inline u8_t _get_num_regions(void)
+
+	/* Populate internal ARM MPU region configuration structure. */
+	region_conf.base = new_region->start;
+	_get_ram_region_attr_by_conf(&region_conf.attr,
```
Sudden change of attributes to configuration which is then converted to the attributes....
I suggest consistent naming.

ioannisg replied:
Not sure exactly what you mean with "consistent naming"
However, you are correct, I don't like having "ram" as an in-fix; this function works with any attribute. So I will update it.

ioannisg replied:
(Addressed by removing "ram" from the function  name)

pizi-nordic replied:
On both layers you have region attribute. However the meaning is different on different layers (it includes size or not). The naming of `_get_ram_region_attr_by_conf()` suggests, that you have an configuration which is converted to attribute. I recommend renaming this function, to avoid a mess in naming. Some thing like get_final_attr to something similar.

ioannisg replied:
Alright, I see what you want. Changing this to `_k_mem_partition_attr_to_mpu_region_attr()`

pizi-nordic commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -78,251 +126,138 @@ void arm_core_mpu_disable(void)
  */
-static inline int _get_region_attr_by_type(arm_mpu_region_attr_t *p_attr,
-	u32_t type, u32_t base, u32_t size)
+void arm_core_mpu_mem_partition_configure(struct k_mem_partition *partition,
```
..._static_region_update()?

ioannisg replied:
This is _not_ for static regions - memory partitions are dynamic.
But, I think, I see your point; this function performs a _configuration update_ of an existing region.

I can add the _update in the function name if you insist, @pizi-nordic

ioannisg replied:
(addressed)

pizi-nordic replied:
Ok.

pizi-nordic commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -78,251 +126,138 @@ void arm_core_mpu_disable(void)
+	 * depending on whether the MPU architecture requires
+	 * non overlapping regions.
+	 */
+#if defined(CONFIG_MPU_REQUIRES_NON_OVERLAPPING_REGIONS)
```
We should not take under account any thread stack and other SW configuration. This layer should get this information from kernel (of course CONFIG_MPU_REQUIRES_NON_OVERLAPPING_REGIONS as HW configuration should be handled here).

ioannisg replied:
Yeah, I see what you mean:
You propose implement the function in arm_core_mpu.c and let the driver (arm_mpu.c, nxp_mpu.c) only return the number of available (HW) regions indices.

Do I understand corretly, @pizi-nordic ?

pizi-nordic replied:
Yes.

ioannisg replied:
Addressed now, thanks.

pizi-nordic commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -78,251 +126,138 @@ void arm_core_mpu_disable(void)
 #endif
 
-	arm_core_mpu_configure(THREAD_STACK_REGION, base, size);
+	return _MPU_MAX_DOMAIN_PARTITIONS_GET(available_regions_num);
```
So where the number of regions should be adjusted to take under account HW and SW config? Above or in this macro?

ioannisg replied:
In this macro. Note that the macro is defined for all the three different driver implementations.

ioannisg replied:
(addressed)

pizi-nordic commented on arch/arm/include/kernel_arch_func.h:
```diff
@@ -90,14 +104,14 @@ _arch_switch_to_main_thread(struct k_thread *main_thread,
 		"mov %%r0, %3 \t\n"
 		"push {r2, lr} \t\n"
-		"blx configure_mpu_stack_guard \t\n"
+		"blx _arch_configure_dynamic_mpu_regions \t\n"
```
What about r3? The clobber list says it is not changed, but here you jumping to complex C code.

ioannisg replied:
True, thanks. I am going to address this as you and the relevant issue suggests.

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -10,105 +10,295 @@
-			    (u32_t)thread->stack_obj;
+#if defined(CONFIG_MPU_STACK_GUARD)
+#define _MAX_DYNAMIC_MPU_REGIONS_NUM \
+	(CONFIG_MAX_DOMAIN_PARTITIONS + 2)
```
Thanks. I am choosing a hybrid solution:
- defining the macro using the IS_ENABLED(.) expressions
- adding supportive inline comment.

ioannisg replied:
(addressed)

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -10,105 +10,295 @@
  *
- * @param thread thread info data structure.
+ * For some MPU architectures, such as the unmodified ARMv8-M MPU,
+ * the function must execute with MPU enabled.
```
We need this for ARMv8-M MPU @pizi-nordic 
The reason is that we need to do run-time partition split and _**to accelerate this**_ we use the ARMv8-M Test-Target (TT) intrinsic;  that requires MPU to be enabled.

So I added this note in the documentation.
Contrary to that, ARMv7-M and NXP MPU do not require MPU to be enabled when executing these functions.

(addressed)

pizi-nordic commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -10,105 +10,295 @@
+	 * the given thread. The array of partitions (along with its
+	 * actual size) will be supplied to the underlying MPU driver.
+	 */
+	struct k_mem_partition dynamic_regions[_MAX_DYNAMIC_MPU_REGIONS_NUM];
```
You have one loop over CONFIG_MAX_DOMAIN_PARTITIONS + priv stack + guard.
We can either
- Try to populate more regions than we have and overflow array (you said that we are safe from this side)
- Skip some things due to lack of memory regions.

My point here (no doubt that well hidden :)) is that we should detect such situations and handle it accordingly. If this is enforced and Kconfig level, then it is OK.

ioannisg replied:
Added ASSERTS, as you suggested :) (you will be able to assess this when I push the revised version)

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -10,105 +10,295 @@
+			 */
+			continue;
+		}
+		arm_core_mpu_mem_partition_configure(&partition, &reset_attr);
```
Thanks for this comment, @pizi-nordic. So, I really thought a lot about this, and I went for this "hybrid" solution of
- re-submitting the map to the driver when we need to actually configure the memory domain on-the-fly (i.e. during current thread's execution)
- just resetting the domain partition / partitions to default ATTRibutes, in case of _destroy or _remove.

The reasons for _not_ doing as you say, are the following:
1. **Performance penalty**: re-submitting the whole map is costly, especially if we only need to remove a single partition (and keep all the remaining ones)
2. We might consider removing this possibility from the Memory Domain API, that is, to be able to _destroy/remove partitions on-the-fly_, and **only** allow it when swapping (discussed with @andrewboie @andyross  and @agross-linaro ).

ioannisg replied:
(addressed)

andrewboie commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -10,105 +10,295 @@
+ * memory area, where dynamic memory regions may be programmed at run-time.
+ */
+#if defined(CONFIG_APP_SHARED_MEM)
+#define _MPU_DYNAMIC_REGIONS_AREA_START ((u32_t)&_app_smem_start)
```
If you want the entire bounds of system RAM, _image_ram_start / _image_ram_end ought to be sufficient instead of all these ifdefs

ioannisg replied:
@andrewboie , I do not want the entire SRAM bounds; I need the boundaries, within which dynamic regions may be defined. Dynamic regions may be defined on Kernel SRAM (e.g. thread stacks), or Application Shared memory (e.g. app partitions).***

Now that I look more about it, the code assumes that these sections are adjacent. I can add a comment on the ARM linker script _not_ to change that. (Will track this externally)

*** Until we remove Application Memory, dynamic (mem domain) partitions may also be defined there. I do take care of this in the code.

ioannisg replied:
(addressed)

andrewboie commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -10,105 +10,294 @@
+#if defined(CONFIG_NOCACHE_MEMORY)
+		{
+		.start = (u32_t)&_nocache_ram_start,
+		.size = (u32_t)&_nocache_ram_end - (u32_t)&_nocache_ram_start,
```
> _start/_end symbols already have char type in linker-defs.h. I don't see why these casts are needed, 

You have to cast because although they are char type, they have no interesting value, it's their *memory address* that is what we want, and that is a pointer, hence the use of &

ioannisg replied:
I suppose, addressed?

andyross replied:
To be clear: the result of subtracting two char* values has type size_t and is in units of bytes, which is AFAICT exactly what you want.  So `(&_foo_end - &foo_start)` should give you an integral value you can use without casting.  But not a big deal.

andrewboie commented on arch/arm/core/cortex_m/mpu/nxp_mpu.c:
```diff
@@ -97,82 +110,256 @@ static void _region_init(u32_t index, u32_t region_base,
+	return mpu_reg_index;
+}
+
+#if 0
```
just delete this #if 0 code if we are not using it

ioannisg replied:
Thanks, this was forgotten. Removed now.

ioannisg replied:
(addressed)

andrewboie commented on arch/arm/core/cortex_m/mpu/nxp_mpu.c:
```diff
@@ -382,14 +540,14 @@ int arm_core_mpu_buffer_validate(void *addr, size_t size, int write)
+	for (r_index = 0U; r_index < _get_num_regions(); r_index++) {
 		if (!_is_enabled_region(r_index) ||
-		    !_is_in_region(r_index, (u32_t)addr, size)) {
+				!_is_in_region(r_index, (u32_t)addr, size)) {
```
typical style is to line up continuations with the opening parenthesis, rather this wasn't changed

ioannisg replied:
Fixing this back, thanks.

(addressed)

andrewboie commented on include/arch/arm/cortex_m/mpu/nxp_mpu.h:
```diff
@@ -86,66 +86,83 @@
 #define ENDADDR_ROUND(x) (x - 0x1F)
 
-#define REGION_USER_MODE_ATTR (MPU_REGION_READ | \
+#define REGION_USER_MODE_ATTR {(MPU_REGION_READ | \
```
why the braces? this doesn't look like a GNU compound expression

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -10,105 +10,294 @@
+#endif
+		dynamic_regions[region_num].start = base;
+		dynamic_regions[region_num].size = size;
+		dynamic_regions[region_num].attr = K_MEM_PARTITION_P_RW_U_RW;
```
Maybe I am wrong, here, @andyross but I think we may need post C99 here, as `dynamic_regions `is an array, i.e. not assignable like this.

ioannisg replied:
(addressed)

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h:
```diff
@@ -35,11 +36,30 @@ static void _region_init(u32_t index, const struct arm_mpu_region *region_conf)
+ * @param part Pointer to the data structure holding the partition
+ *             information (must be valid).
+ * */
+static int _mpu_partition_is_sane(const struct k_mem_partition *part)
```
(addressed)

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -10,105 +10,294 @@
+			if (partition.size == 0) {
+				/* Zero size indicates a non-existing
+				 * memory partition.
+				 */
```
This is not a dumb question Andy :)

However: by-design (please see in `kernel/mem_domain.c`) a zero-size denotes a free partition slot:

A partition is removed by zero-ing its size:

```
	/* A zero-sized partition denotes it's a free partition */
	domain->partitions[p_idx].size = 0;
```
The earlier version of the ARM MPU driver has also been handling mem-domain operations assuming this design principle.

(Addressed)

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -10,105 +10,294 @@
+			region_num++;
+			num_partitions--;
+			if (num_partitions == 0) {
+				break;
```
So:
- `i `iterates on the input array of partition items (some elements of this array may be null entries)
- `region_num `increments and populates the local array of dynamic memory map. We can't really use i for this, because of the possible null entries on the input array.

>early exit above (that maybe could be eliminated?)

Actually, I would like to keep it, so I do not run the FOR loop until `CONFIG_MAX_DOMAIN_PARTITIONS`, in case all non-null elements are already exhausted.

The num_partitions (the variable and this check) is the price we have to pay, to be able to terminate early the loop (this makes some sense for MPU hardware with quite many HW regions, e.g. 32 but only a few actually configured partitions)

ioannisg replied:
(addressed)

ioannisg commented on arch/arm/core/cortex_m/mpu/nxp_mpu.c:
```diff
@@ -401,8 +557,40 @@ int arm_core_mpu_buffer_validate(void *addr, size_t size, int write)
+		LOG_ERR("Configuring static MPU regions failed");
+		__ASSERT(0, "Configuring %u static MPU regions failed\n",
+			regions_num);
+	}
```
OK, I am to blame for the confusion here, @andyross .
So: in this particular case **we shall not be able to continue**, if we can't configure the Static regions. So the ASSERT is required. Perhaps remove the LOG_ERR, then?

ioannisg replied:
REmoved the LOG_ERR.

andyross replied:
Seems like you removed the wrong one, though.   The error is still happening at runtime and the code in production mode is still going to stumble ahead despite having encountered a fatal/unforseen condition!

What that assert means is that _mpu_configure_static_mpu_regions() **must never fail**.  Well... then why are you bothering to check its return value?

ioannisg commented on arch/arm/core/cortex_m/mpu/nxp_mpu.c:
```diff
@@ -401,8 +557,40 @@ int arm_core_mpu_buffer_validate(void *addr, size_t size, int write)
+		LOG_ERR("Configuring dynamic MPU regions failed");
+		__ASSERT(0, "Configuring %u dynamic MPU regions failed\n",
+			regions_num);
+	}
```
ASSERT (similar to the case above)

(addressed)
PS: Not being able to configure memory map shall be an unrecoverable error, and for that we need to have ASSERTS during debug/develop phase,  agree @andrewboie ?

ioannisg commented on arch/arm/core/cortex_m/mpu/nxp_mpu.c:
```diff
@@ -411,73 +599,48 @@ int arm_core_mpu_buffer_validate(void *addr, size_t size, int write)
+			mpu_config.num_regions,
+			_get_num_regions()
+		);
+		return -1;
```
Alright, here, we have an issue:

the mpu_init function returns -1 on error, **but no-one reads this signal**, so we, traditionally have an ASSERT to catch configuration error at Boot time. So, I think this can stay....

ioannisg replied:
(Addressed)

ioannisg commented:
@andyross @andrewboie I have addressed your comments - appreciate if you could re-visit the review.

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -10,105 +10,294 @@
+				partition.start, partition.size);
+			dynamic_regions[region_num].start = partition.start;
+			dynamic_regions[region_num].size = partition.size;
+			dynamic_regions[region_num].attr = partition.attr;
```
Yes. Fixing this as you propose.

Note, however, that this cannot be applied exactly like this, in the following assignments of `dynamic_regions[region_num]
`

(addressed)

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -10,105 +10,294 @@
+	    (u32_t)thread->stack_obj;
+#else
+	u32_t guard_start = thread->stack_info.start;
+#endif
```
Ah, I think I wouldn't do it - especially since arm_core_mpu.c does not contain any other private functions, at all. So readers should "train" to see all implementation in the API functions :)

ioannisg replied:
Unless you insist, of course..

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -10,105 +10,294 @@
+	 * This triggers re-programming of the entire dynamic
+	 * memory map.
+	 */
+	_arch_configure_dynamic_mpu_regions(thread);
```
You are right @andyross, this is a bit confusing, and, _slightly_ problematic, but I believe we can live with that.

In brief:
- `_arch_mem_domain_configure()` is an existing API from `kernel/mem_domain.h`. As such, I want to preserve it as is, in this re-work of the ARM MPU.
- `_arch_configure_dynamic_mpu_regions ()` is the new, single, universal API function for re-programming the Memory map at run-time (user-mode enter and context-switch). As such, `_arch_configure_dynamic_mpu_regions ()` also programs the (app) memory domains.

Now: app memory domains may _also_ be programmed at run-time at arbitrary places (i.e. any privileged thread, or IRQ is allowed do it). I just felt that I could **_reuse_** the existing function of programming the dynamic memory map, instead of adding a new ARM Core MPU function that programs only the memory partitions.

The comment:
```
/*
 * This triggers re-programming of the entire dynamic
 * memory map.
 */

indicates my true intentions.

>Can't you just use the old function name universally?

This will be done in the future, actually: I am going to use the `_arch_configure_dynamic_mpu_regions()` even inside the kernel/mem_domain.c.
But not in this PR :)
```

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -10,105 +10,294 @@
+			/* Zero size indicates a non-existing
+			 * memory partition.
+			 */
+			continue;
```
I believe this is answered in the other place :)

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -49,6 +45,58 @@ static inline u8_t _get_num_regions(void)
+
+		/* No available MPU region index. */
+		__ASSERT(0, "Failed to allocate new MPU region %u\n", index);
+		return -EINVAL;
```
Thanks, the ASSERT will be removed.

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -16,17 +16,13 @@
+ * have been reserved by the MPU driver to program the static (fixed) memory
+ * regions. The variable is meant to be used at run-time to derive the number
+ * of HW MPU region indices available for programming dynamic memory regions.
+ */
```
Agree, for the most part.
I will keep the first sentence; not everyone understands, easily, what "static memory regions" are.
I will remove the second sentence.

If you still insist, I can remove it all, no strong opinion about that.

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h:
```diff
@@ -251,9 +232,121 @@ static inline int _mpu_buffer_validate(void *addr, size_t size, int write)
+		/* Non-empty region. */
+
+		if (do_sanity_check &&
+				(!_mpu_partition_is_sane(&regions[i]))) {
```
Not exactly - I might have failed to explain this properly, Andy:

Sanity check is _required_ for partitions configured at run-time; if they are invalid, the MPU behavior is undefined. The boolean variable is used for telling the function not do do sanity check - this is used when configuring static regions, whose parameters are set at compile time, and we are confident that they are "sane".

It does not have to do with ASSERTS being on or off.

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h:
```diff
@@ -251,9 +232,121 @@ static inline int _mpu_buffer_validate(void *addr, size_t size, int write)
+
+		if (do_sanity_check &&
+				(!_mpu_partition_is_sane(&regions[i]))) {
+			__ASSERT(0, "Partition %u: sanity check failed.", i);
```
Right. The error is, anyways, propagated at run-time, through the return value.

ioannisg replied:
Removed the ASSERT - replaced with error reporting.

andyross commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -78,251 +125,113 @@ void arm_core_mpu_disable(void)
-	} else {
-		_disable_region(region_index + part_index);
+		__ASSERT(0, "Marking %u areas for dynamic regions failed\n",
+			dyn_region_areas_num);
```
Sorry, this is another spot I think I missed earlier.  On a development build this will fail synchronously right here, but on a production build it will plod along having disabled the region.  The intent of __ASSERT() is that this else() clause **should never execute because if it did that would be a bug**, so why does it contain code?

ioannisg replied:
I suppose this comment can be resolved.

andyross commented on arch/arm/core/cortex_m/mpu/arm_mpu_v8_internal.h:
```diff
@@ -145,6 +243,255 @@ static inline int _mpu_buffer_validate(void *addr, size_t size, int write)
+		if ((u_reg_index == -EINVAL) ||
+			(u_reg_index > (reg_index - 1))) {
+			__ASSERT(0,	"Invalid underlying region index %u",
+				u_reg_index);
```
LOG_ERR() here.

andyross reviewed (approved):
Still really not liking the "use ASSERT(0) for reporting runtime errors" trick, but it seems like all the usage here is no longer polluted with other runtime behavior.

Not worth continuing the argument really, but it.... does kind of have a bad code smell.  I'd worry that some of those "fatal errors that should never happen" would over time become "fatal errors we EXPECT not to happen" and then become real/potential runtime errors.  This scheme would be unable to detect them or test for them except by hitting them in a production build.

ioannisg commented:
>

Then change this to an _ASSERT(!IS_ENABLED(CONFIG_MPU_REQUIRES_NON_OVERLAPPING_REGIONS), "") maybe? That would better express this as a configuration error and not a runtime one.

Or just printk, honestly. No shame in printk.

@andyross, I agree, that this could be LOG_ERR, but I left the ASSERT since this is in accordance with the rest of the code in this kernel module (`kernel/mem_domain.c`). Also, the file does not seem to have migrated to the new logger.

I suppose that once this is done, one can replace all ASSERTS with LOG_ERR, at least in those cases where we return an error signal...

ioannisg commented:
>

Seems like you removed the wrong one, though. The error is still happening at runtime and the code in production mode is still going to stumble ahead despite having encountered a fatal/unforseen condition!

What that assert means is that _mpu_configure_static_mpu_regions() must never fail. Well... then why are you bothering to check its return value?

You are right on this one, too, @andyross however:
the reason for keeping the _ASSERT is that there is still no infrastructure for core arch/arm functions called in swap, userspace-enter, etc. to check error values. Once we will be able to do that, `arm_core_mpu_configure_static_mpu_regions `will be made to return int, and I will remove the ASSERT. I keep it for now as there is no other way to track the error in dev stage (and halt the execution)

andrewboie commented on arch/Kconfig:
```diff
@@ -413,6 +413,12 @@ config MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT
 	  and size for MPU regions.
 
+config MPU_REQUIRES_NON_OVERLAPPING_REGIONS
+	bool
```
@ruuddw likely useful for ARC as well, the V3 ARC MPU also doesn't allow overlaps

tyrantlizard77 reviewed (approved):
I only have superficial knowledge of the ARM MPU internals, but this looks sane to me in its current form.

agansari commented on arch/arm/core/cortex_m/mpu/nxp_mpu.c:
```diff
@@ -97,95 +111,251 @@ static void _region_init(u32_t index, u32_t region_base,
+	if (index > (_get_num_regions() - 1)) {
 
+		/* No available MPU region index. */
+		return -EINVAL;
```
Add a warning message as in arm_mpu.c
LOG_ERR("Failed to allocate new MPU region %u\n", index);

ioannisg replied:
Thanks. Will do.

agansari commented on arch/arm/core/cortex_m/mpu/nxp_mpu.c:
```diff
@@ -195,7 +365,6 @@ void arm_core_mpu_enable(void)
 	/* Enable MPU */
 	SYSMPU->CESR |= SYSMPU_CESR_VLD_MASK;
-
 }
```
add __DSB(); here and remove the barriers at the end of nxp_mpu_init(), this way barriers are always carried on when enabling/disabling mpu not only at mpu initialization.

ioannisg replied:
Yeap, will do. ISB?

agansari commented on arch/arm/core/cortex_m/mpu/nxp_mpu.c:
```diff
@@ -209,165 +378,110 @@ void arm_core_mpu_disable(void)
 	SYSMPU->CESR |=  SYSMPU_CESR_SPERR_MASK;
```
Add __DSB(); before /* Disable MPU */
Remove the barriers at the end of nxp_mpu_init(), this way barriers are always carried on when enabling/disabling mpu not only at mpu initialization.

ioannisg replied:
Sure.

ioannisg replied:
Added

agansari commented on arch/arm/core/cortex_m/mpu/nxp_mpu.c:
```diff
@@ -411,80 +555,55 @@ int arm_core_mpu_buffer_validate(void *addr, size_t size, int write)
-#endif
 	/* Make sure that all the registers are set before proceeding */
 	__DSB();
 	__ISB();
```
Remove the barriers at the end of nxp_mpu_init().

ioannisg replied:
Done.

agansari reviewed (changes requested):
The changes look okay, NXP's MPU implementation is similar to ARM. I have a few minor requests.
I've tested this patch on K64F, tests passed.

aurel32 commented on arch/arm/include/kernel_arch_func.h:
```diff
@@ -42,6 +46,16 @@ _arch_switch_to_main_thread(struct k_thread *main_thread,
+	 *
+	 * This function is invoked once, upon system initialization.
+	 */
+	_arch_configure_static_mpu_regions();
```
I think that doing that configuring those region in `_arch_switch_to_main_thread()` is a bit late, at least for the `__nocache` region, which has to be setup when the drivers (currently GMAC and MCUX ethernet drivers) are initialized. IIUC the MPU initialization is currently done in PRE_KERNEL_1.

ioannisg replied:
Thanks, this is interesting. So, the MPU driver is initialized in PRE_KERNEL_1, correct.

Could you elaborate a bit why the __no_cache region can't wait until kernel/init, which comes before we jump to the _main() (Main thread)? This comes before POST_KERNEL, as far as I can see, anyways.

Otherwise, sure, we could always move this call into the driver itself: Static regions _by design_ do not depend on thread(s)/kernel. But if no big deal, I would prefer to keep the call where it is now :)

ioannisg commented:
> 
> 
> The changes look okay, NXP's MPU implementation is similar to ARM. I have a few minor requests.
> I've tested this patch on K64F, tests passed.

@agansari Thanks for feedback & testing on NXP platform. Your comments are now addressed; pls, revisit. :)

aurel32 commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -10,11 +10,99 @@
+		{
+		.start = (u32_t)&__app_ram_start,
+		.size = (u32_t)&__app_ram_end - (u32_t)&__app_ram_start,
+		.attr = K_MEM_PARTITION_P_RW_U_RW,
```
The `K_MEM_PARTITION*` defines do not include the cacheability attributes, so we are effectively loosing the TEX, C, B and S bits compare to the previous `_get_mpu_ram_region_attr`.

ioannisg replied:
@aurel, the K_MEM_PARTITION_ defines _do_ include the cacheability attributes. When they don't default Cache is assumed. And we also have the _NOCACHE for the __no_cache memory regions.

aurel32 replied:
I don't see where the cache attributes are added in ` include/arch/arm/cortex_m/mpu/arm_mpu_v7m.h.` I wanted to give it a try to check and check the values at runtime, but it doesn't even compile:

```
In file included from /home/aurel32/git/zephyrproject/zephyr/include/arch/arm/cortex_m/mpu/arm_mpu.h:13:0,
                 from /home/aurel32/git/zephyrproject/zephyr/include/arch/arm/arch.h:257,
                 from /home/aurel32/git/zephyrproject/zephyr/include/arch/cpu.h:17,
                 from /home/aurel32/git/zephyrproject/zephyr/include/kernel_includes.h:34,
                 from /home/aurel32/git/zephyrproject/zephyr/include/kernel.h:17,
                 from /home/aurel32/git/zephyrproject/zephyr/include/device.h:11,
                 from /home/aurel32/git/zephyrproject/zephyr/arch/arm/core/cortex_m/mpu/arm_core_mpu.c:7:
/home/aurel32/git/zephyrproject/zephyr/arch/arm/core/cortex_m/mpu/arm_core_mpu.c: In function '_arch_configure_static_mpu_regions':
/home/aurel32/git/zephyrproject/zephyr/include/arch/arm/cortex_m/mpu/arm_mpu_v7m.h:219:2: error: invalid operands to binary | (have 'k_mem_partition_attr_t {aka struct <anonymous>}' and 'long unsigned int')
  | NORMAL_OUTER_INNER_NON_CACHEABLE_SHAREABLE)})
  ^
/home/aurel32/git/zephyrproject/zephyr/arch/arm/core/cortex_m/mpu/arm_core_mpu.c:86:11: note: in expansion of macro 'K_MEM_PARTITION_P_RW_U_RW_NOCACHE'
   .attr = K_MEM_PARTITION_P_RW_U_RW_NOCACHE,
           ^
```

aurel32 replied:
I workarounded this issue and then compared the configuration of the application memory:

On master:
`[00:00:00.000,000] <dbg> mpu._region_init: [4] 0x20400000 0x130b0016`

With PR#12817:
`[00:00:00.000,013] <dbg> mpu._region_init: [3] 0x20400000 0x13000016`

It means TEX, C and B went from 1 to 0.

aurel32 commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -10,11 +10,99 @@
+		{
+		.start = (u32_t)&_nocache_ram_start,
+		.size = (u32_t)&_nocache_ram_end - (u32_t)&_nocache_ram_start,
+		.attr = K_MEM_PARTITION_P_RW_U_RW_NOCACHE,
```
OTOH, this one looks good. This probably means we want to introduce `K_MEM_PARTITION_P_RW_U_RW_CACHE`. Alternatively the oring might be done in this file instead of adding too many defines.

ioannisg replied:
@aurel32 _CACHE is  (I assume) by default, pls consult the public arm_mpu.vx.h headers :)

agross-linaro commented:
Tested on musca-a.  Userspace and stack protection tests passed.  Had 2 failures in the tests/kernel/mem_protect/mem_protect:

starting test - test_mem_domain_add_partitions_invalid

    Assertion failed at /work/zep/tests/kernel/mem_protect/mem_protect/src/mem_domain.c:333: mem_domain_for_user_tc3: (Reached unreachable code)
Fault didn't occur when we accessed a unassigned memory domain.
FAIL - test_mem_domain_add_partitions_invalid

starting test - test_thread_has_residual_permissions

    Assertion failed at /work/zep/tests/kernel/mem_protect/mem_protect/src/kobject.c:379: kobject_user_2_test9: (Reached unreachable code)
Failed to clear premission on a deleted thread
FAIL - test_thread_has_residual_permissions

These *DO NOT* fail on frdm_k64f

aurel32 reviewed (changes requested):
I have finished reviewing this PR. It looks overall fine, but there is a confusion between `k_mem_partition.attr` and the `RASR` register of the MPU. `k_mem_partition.attr` used to represent only the permission part of the register (`XN` and `AP` field), while with this PR it also includes the cache attributes. As `k_mem_partition` is public and used to define a memory partition, this is actually an API change.

Now this can probably be fixed by updating all the `K_MEM_PARTITION_*` defines. It also means that the cacheability is now configurable per partition, but I don't think it's really an issue, it's probably more a bonus.

I have only lightly tested this PR due to the above issues, that said it seems to work fine.

ioannisg commented:
> 
> 
> I have finished reviewing this PR. It looks overall fine, but there is a confusion between `k_mem_partition.attr` and the `RASR` register of the MPU. `k_mem_partition.attr` used to represent only the permission part of the register (`XN` and `AP` field), while with this PR it also includes the cache attributes. As `k_mem_partition` is public and used to define a memory partition, this is actually an API change.
> 
> Now this can probably be fixed by updating all the `K_MEM_PARTITION_*` defines. It also means that the cacheability is now configurable per partition, but I don't think it's really an issue, it's probably more a bonus.
> 
> I have only lightly tested this PR due to the above issues, that said it seems to work fine.

@aurel32 thanks a lot for the feedback; i realized now missed including a commit to the PR which adds cacheability to the standard K_MEM_PARTITION_xxx macros. I will have to do this and ask you to look back at the PR.

ioannisg committed:
arch: arm: mpu: update K_MEM_PARTITION_ macros for ARMv7-M

The ARM core MPU API now uses solely k_mem_partition_attr_t
objects to represent memory region attributes. The objects
now include all attribution properties (including cache-
ability and share-ability). This commit updates the macro
definitions to comply with the new ARM Core MPU standard.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
> 
> 
> I have finished reviewing this PR. It looks overall fine, but there is a confusion between `k_mem_partition.attr` and the `RASR` register of the MPU. `k_mem_partition.attr` used to represent only the permission part of the register (`XN` and `AP` field), while with this PR it also includes the cache attributes. As `k_mem_partition` is public and used to define a memory partition, this is actually an API change.
> 
> Now this can probably be fixed by updating all the `K_MEM_PARTITION_*` defines. It also means that the cacheability is now configurable per partition, but I don't think it's really an issue, it's probably more a bonus.
> 
> I have only lightly tested this PR due to the above issues, that said it seems to work fine.

@aurel32 thanks a lot for looking at the PR; I have pushed a (_forgotten_) commit that updates the K_MEM_PARTITION_xx macros according to the new design. Please, re-visit :)

aurel32 commented:
> Fixes #12548 (to check with @aurel32)

I confirm this fixes the issue.

aurel32 commented on arch/arm/core/cortex_m/mpu/arm_core_mpu.c:
```diff
@@ -10,11 +10,99 @@
+		{
+		.start = (u32_t)&_nocache_ram_start,
+		.size = (u32_t)&_nocache_ram_end - (u32_t)&_nocache_ram_start,
+		.attr = K_MEM_PARTITION_P_RW_U_RW_NOCACHE,
```
This should actually be changed into `K_MEM_PARTITION_P_RW_U_NA_NOCACHE` like it is in the existing implementation. So far there is no reason to give the user threads access to the __nocache region. This region is supposed to be used in drivers to interface with hardware.

ioannisg replied:
Fixed that, now, @aurel32 , thanx

aurel32 commented:
> @aurel32 thanks a lot for looking at the PR; I have pushed a (_forgotten_) commit that updates the K_MEM_PARTITION_xx macros according to the new design. Please, re-visit :)

Thanks, with the additional commit it looks good to me, beside the trivial change needed for the permissions of the __nocache region. I have tested it successfully with various apps. I have also launched all the kernel tests on a SAM E70 Xplained board and they all pass. I haven't tried on another board though.

ioannisg commented:
> 
> 
> > @aurel32 thanks a lot for looking at the PR; I have pushed a (_forgotten_) commit that updates the K_MEM_PARTITION_xx macros according to the new design. Please, re-visit :)
> 
> Thanks, with the additional commit it looks good to me, beside the trivial change needed for the permissions of the __nocache region. I have tested it successfully with various apps. I have also launched all the kernel tests on a SAM E70 Xplained board and they all pass. I haven't tried on another board though.

the change is perhaps trivial, but essential, thanks for catching this.

could you update the PR description and add the board in the list of the ones where this re-work was tested?

agross-oss reviewed (approved):
There appears to be an issue on musca, but I'm willing to merge now and work that as a bug.  The musca is such low volume and only 3-4 people even have one.
